Page description

KalobTaulien opened the pull request:
* Do the tests still pass? ✅
* Does the code comply with the style guide? ✅
* Have you added tests to cover the new/fixed behaviour? ✅
* Has the documentation been updated accordingly? ✅

Added a "help text"-style optional attribute to the Page model that will display the Page's helpful text in three different locations in the Wagtail admin. Previews below.

**Note:** I did _not_ add any frontend styling. I'd be happy for some constructive feedback on the design implementation and would be happy to open another PR for better design/frontend implementation. 

<img width="402" alt="Screen Shot 2019-10-07 at 7 19 07 PM" src="https://user-images.githubusercontent.com/4743971/66362293-61879e00-e93f-11e9-9adc-5fa17f4bf408.png">

tomdyson commented:
Thanks @KalobTaulien! 

Should we consider using [Meta options](https://docs.djangoproject.com/en/3.0/topics/db/models/#meta-options) for optional non-content attributes like this one?

> Model metadata is “anything that’s not a field”, such as ordering options (ordering), database table name (db_table), or human-readable singular and plural names (verbose_name and verbose_name_plural). None are required, and adding class Meta to a model is completely optional.

gasman commented:
I think we tried using Meta for this sort of thing in the early days of Wagtail, and found that Django made it hard to add extra properties beyond the built-in set. Maybe that's been improved now though...

jacobtoppm commented on wagtail/admin/templates/wagtailadmin/pages/create.html:
```diff
@@ -13,7 +13,11 @@
             <div class="left col9 header-title">
                 <h1 class="icon icon-doc-empty-inverse">
-                {% trans 'New' %} <span>{{ content_type.model_class.get_verbose_name }}</span></h1>
+                    {% trans 'New' %} <span>{{ content_type.model_class.get_verbose_name }}</span>
```
My concern here is for screen readers, though I'm not sure what the solution is. I tested this briefly in Orca in Firefox, and found that the small tag did not cause a break - there's no separation between the heading and subheading. While this isn't a huge issue, it could lead to long and confusing headings for screen reader users. Perhaps this would be better outside the h1 to separate it, or by using something else to denote the heading/subheading divide?

zerolab replied:
Semantically, the description does not belong in the heading, so 👍for moving it outside.
This may also move the description under "New/Editing <Page type>" making it cleaner.

thibaudcolas replied:
I would suggest moving it out because I presume the text could get relatively lengthy for real-world descriptions.

If you do want to keep it in, you can add a screen-reader-only full stop `<span class="visuallyhidden">.</span>` in-between the title and description so there is a pause when announcing the text. In that case it would be nice for the `<title>` and `<h1>` to match.

jacobtoppm commented on wagtail/admin/templates/wagtailadmin/pages/edit.html:
```diff
@@ -13,7 +13,11 @@
             <div class="left col9 header-title">
                 <h1 class="icon icon-doc-empty-inverse">
-                {% blocktrans with title=page.get_admin_display_title page_type=content_type.model_class.get_verbose_name %}Editing {{ page_type }} <span>{{ title }}</span>{% endblocktrans %}</h1>
+                    {% blocktrans with title=page.get_admin_display_title page_type=content_type.model_class.get_verbose_name %}Editing {{ page_type }} <span>{{ title }}</span>{% endblocktrans %}
```
This combination of page type + page name + help text could get pretty long, and doesn't look great when it wraps as far as I can tell when testing quickly - is there any way we could look at the wrapping here, or put the help text below instead?

KalobTaulien replied:
Hey @jacobtoppm I'm going to remove this until the new design work is in core as to prevent design conflicts.

KalobTaulien commented:
I rebased this off master today. 

I've removed the edit view changes since there's lots of great design work coming down the pipeline soon. 

I only added the one assert in the existing add_supage test, not sure if it requires more or not. Happy to write more if that's what we think is best. 

And I've pinged @thibaudcolas for his thoughts on accessibility in the /add_subpage/ template for the last comment. 

@gasman should I look into moving this into the Meta class as Tom had mentioned? 

Thanks for the feedback everybody!

thibaudcolas commented on wagtail/core/models.py:
```diff
@@ -1073,6 +1073,13 @@ def get_verbose_name(cls):
+    @classmethod
+    def get_page_description(cls):
+        """
+        Returns a page description if it's not set. eg "A multi-purpose web page".
```
Typo?

```suggestion
        Returns a page description if it's set. eg "A multi-purpose web page".
```

thibaudcolas replied:
Bonus point for documentation – most content guidelines focused on readability suggest avoiding "eg" or other latin abbreviation and going for more plain language like "For example" instead.

thibaudcolas commented on wagtail/admin/templates/wagtailadmin/pages/create.html:
```diff
@@ -6,13 +6,18 @@
+                    {% trans 'New' %} <span>{{ content_type.model_class.get_verbose_name }}</span>
+                </h1>
+                {% if content_type.model_class.get_page_description %}
+                    <h3>{{ content_type.model_class.get_page_description }}</h3>
```
This would be better as a `p` or un-semantic `div`. There shouldn’t be two headings one after another like this.

thibaudcolas commented on wagtail/admin/templates/wagtailadmin/pages/add_subpage.html:
```diff
@@ -13,14 +13,19 @@
+                                <a href="{% url 'wagtailadmin_pages:add' app_label model_name parent_page.id %}{% if next %}?next={{ next }}{% endif %}" class="icon icon-plus-inverse icon-larger">
+                                    {{ verbose_name }}
+                                    {% if page_description %}
+                                        <div><small>{{ page_description }}</small></div>
```
Sorry I didn’t catch this sooner – could you add a:

```suggestion
                                        <div><span class="visuallyhidden">.</span><small>{{ page_description }}</small></div>
```

This will have screen readers have a pause between announcing the name and the description.

gasman commented:
> @gasman should I look into moving this into the Meta class as Tom had mentioned?

Sure - definitely worth looking, although I seem to remember that last time I checked, Django didn't let you add arbitrary new options to Meta, so don't worry too much if it can't be done.

kmagee123 commented:
Hi @KalobTaulien here is the design that was requested for this feature. Tom Dyson has had a quick look at this yesterday and gave it the go-ahead. Let me know if you have any question. 

[]
![Artboard (1)](https://user-images.githubusercontent.com/60231423/89995949-1ab2b500-dc82-11ea-8ef1-75de37b1a386.png)
(url)

kmagee123 commented:
Hi @KalobTaulien here is the design that was requested for this feature. Tom Dyson has had a quick look at this yesterday and gave it the go-ahead. Let me know if you have any question.

![Wagtail design](https://user-images.githubusercontent.com/60231423/89996548-d673e480-dc82-11ea-9261-7e4ecbe6ec42.png)

allcaps commented:
Only `DEFAULT_NAMES` are allowed on Meta.
https://github.com/django/django/blob/master/django/db/models/options.py#L198

This:
```
class FooPage(Page):
    class Meta:
        description = "foo"
```
Gives an error on startup:
```
TypeError: 'class Meta' got invalid attribute(s): description
```

We can extend `DEFAULT_NAMES`, but I prefer a method. 

A method allows easier interaction with the rest of the application which can be useful. Eg: Display the page usage count in the description, or get some other dynamic data.

allcaps commented on wagtail/admin/templates/wagtailadmin/pages/create.html:
```diff
@@ -6,13 +6,18 @@
+                    {% trans 'New' %} <span>{{ content_type.model_class.get_verbose_name }}</span>
+                </h1>
+                {% if content_type.model_class.get_page_description %}
+                    <p>{{ content_type.model_class.get_page_description }}</p>
```
Can we change this `p` to `div`? One might return a safe string from `get_page_description` with html structure. This allows you to display thumbnail/previews etc.

allcaps reviewed (approved):
LGTM.

Small thing, the `get_page_description` might also return a HTML structure. So I prefer wrapping the content in a div.

thibaudcolas commented:
This seems ready to go – I’ll review the conflicts, try to resolve them, and merge. We are going to be redesigning this page in the near future, so I think the design by Kyle can already be considered outdated unfortunately.

zerolab commented:
Superseded by #8065 which fixes merge conflicts
